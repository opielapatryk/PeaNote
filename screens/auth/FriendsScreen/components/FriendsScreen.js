import { Text,View, Pressable, TextInput,FlatList } from 'react-native'
import React, { useEffect } from 'react'
import {sendFriendRequest} from '../functions/sendFriendRequest'
import {styles} from '../../../../assets/styles/styles'
import { useDispatch, useSelector } from 'react-redux';
import {setEmail,setMessage} from '../../../../store/login/loginSlice'
import { useFocusEffect } from '@react-navigation/native';
import { loadUser } from '../functions/loadUser';
import { cleanStoreFriends } from '../../../../store/friends/friendsSlice';
import { useSafeAreaInsets } from 'react-native-safe-area-context';

export const FriendsScreen = ({ navigation }) => {
  const {friends} = useSelector((state) => state.friends);
  const {message,email} = useSelector((state) => state.login)
  const dispatch = useDispatch()

  useFocusEffect(
    React.useCallback(() => {
      loadUser(dispatch)
      return ()=>{
        dispatch(setMessage(''))
        dispatch(setEmail(''))
      }
    }, [])
  );

  useEffect(()=>{
    loadUser(dispatch)
  },[])

  const RenderFriendsMemoized = React.memo(({ item }) => {
    return (
      <Pressable onPress={() =>navigation.navigate('FriendsBoard', {name:item.username, friendEmail: item.email, oldnickname:item.nickname})} style={styles.friendsList}><Text style={styles.firendListText}>{item.nickname?item.nickname:item.username}</Text></Pressable>
    );
  });

  const renderFriends = ({ item }) => <RenderFriendsMemoized item={item} />;

  const insets = useSafeAreaInsets();

  return (
    <View style={[styles.board,{
      paddingTop: insets.top,
      flex: 1,
      backgroundColor:'white',
    }]}>
      <Pressable style={styles.friendsHeaderRequest} onPress={()=>navigation.navigate('Requests')}><Text style={styles.removeFriendText}>REQUESTS</Text></Pressable>

      <TextInput style={styles.friendsTextInput} placeholder={message?message:'SEARCH BY EMAIL OR USERNAME'} onChangeText={text => dispatch(setEmail(text))} value={email} autoCorrect={false}/>

      <Pressable style={styles.friendsHeaderRequest} onPress={()=>sendFriendRequest(dispatch,email,navigation)}><Text style={styles.removeFriendText}>SEARCH</Text></Pressable>

      <FlatList data={friends} renderItem={({item})=>renderFriends({item})} keyExtractor={(friend) => friend.id}/>
    </View>
  );
}

export default FriendsScreen