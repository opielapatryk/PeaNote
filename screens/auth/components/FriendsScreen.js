import { Text,ScrollView, Button, TextInput } from 'react-native'
import React, {useState,useEffect} from 'react'
import { useFocusEffect } from '@react-navigation/native';
import {searchNewFriend,addNewFriend,loadUser,getUserId} from '../logic/apiFriendsScreen'

export const FriendsScreen = ({ navigation }) => {
  const [friends,setFriends] = useState([])
  const [newFriendEmail, setNewFriendEmail] = useState()
  const [buttonTitle,setButtonTitle] = useState('')
  const [newFriendID, setNewFriendID] = useState(null)
  const [message,setMessage] = useState('')
  const [doesEmailExist, setdoesEmailExist] = useState(false)
  const [list,setList] = useState([])
  const [firstRender, setFirstRender] = useState(true)

  useEffect(() => {
    getUserId(doesEmailExist,firstRender,setMessage,setButtonTitle,setFirstRender,newFriendID,list,newFriendEmail)
  }, [newFriendID]);

  useFocusEffect(
    React.useCallback(() => {
      loadUser(setFriends)
    }, [])
  );

  return (
    <ScrollView>
      <Button title='REQUESTS' onPress={()=>navigation.navigate('Requests')}/>
      <TextInput placeholder='Insert friend email' onChangeText={(newFriendEmail) => setNewFriendEmail(newFriendEmail)} value={newFriendEmail}/>
      <Button title='SEARCH NEW FRIEND' onPress={()=>searchNewFriend(newFriendEmail,setNewFriendID,setdoesEmailExist,setList)}/>
      <Text>{message}</Text>
      <Button title={buttonTitle} onPress={()=>addNewFriend(setFriends,newFriendID,friends)}/>
      <Text>Friends:</Text>
      <ScrollView>
        {friends.map((friend)=>(
          <Button key={friend.id} title={friend.email} onPress={() => navigation.navigate('FriendsBoard', { friendId: friend.id, friendName: friend.first_name })}/>
        ))}
      </ScrollView>
    </ScrollView>
  );
}

export default FriendsScreen