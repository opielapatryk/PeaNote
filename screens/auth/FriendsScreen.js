import { Text,ScrollView, Button, TextInput } from 'react-native'
import React, {useState,useEffect} from 'react'
import axios from 'axios'
import * as SecureStore from 'expo-secure-store';
import { useFocusEffect } from '@react-navigation/native';
import { useDispatch, useSelector} from 'react-redux';
import {removeNote} from '../../store/boardSlice';
import {userLink,usersLink} from '../../components/Constants'

export const FriendsScreen = ({ navigation }) => {
    const { notes } = useSelector((state)=>state.board)
    const dispatch = useDispatch()
    const [friends,setFriends] = useState([])
    const [newFriendEmail, setNewFriendEmail] = useState()
    const [buttonTitle,setButtonTitle] = useState('')
    const [newFriendID, setNewFriendID] = useState(null)
    const [message,setMessage] = useState('')
    const [doesEmailExist, setdoesEmailExist] = useState(false)
    const [list,setList] = useState([])
    const [firstRender, setFirstRender] = useState(true)

    useFocusEffect(
      React.useCallback(() => {
        const loadUser = async()=>{
          try{
            currentUserId = await SecureStore.getItemAsync('userId');
  
            const result = await axios.get(userLink(currentUserId))
  
            const friendsRequests = result.data.friends.map(url =>
              axios.get(url)
              .then(response => response.data)
            );
          
            Promise.all(friendsRequests)
              .then(friendsData => {
                  setFriends(friendsData);
              })
              .catch(error => {
                console.error('Error fetching friends:', error);
            });
            
            return result
          }catch(e){
            console.log(e.message)
          }
        }
        loadUser()

      }, [])
    );

    useEffect(()=>{
      const loadUser = async()=>{
        try{
          currentUserId = await SecureStore.getItemAsync('userId');

          const result = await axios.get(userLink(currentUserId))

          const friendsRequests = result.data.friends.map(url =>
            axios.get(url)
            .then(response => response.data)
          );
        
          Promise.all(friendsRequests)
            .then(friendsData => {
              if (JSON.stringify(friends) !== JSON.stringify(friendsData)) {
                setFriends(friendsData);
              }
            })
            .catch(error => {
              console.error('Error fetching friends:', error);
          });
          
          return result
        }catch(e){
          console.log(e.message)
        }
      }
      loadUser()
    },[friends])

    useEffect(() => {
      const getUserId = async()=>{
        currentUserId = await SecureStore.getItemAsync('userId');

        if(doesEmailExist === false){
          if(!firstRender){
            setMessage('This user does not exist!..')
            setButtonTitle('')
          }
          setFirstRender(false)
        }else{
          if(newFriendID == currentUserId){
            setMessage('You cannot add to friends yourself!..')
            setButtonTitle('')
          }else{
            if(list.includes(userLink(newFriendID))){
              setMessage('You are friends already!..')
              setButtonTitle('')
            }else{
              setMessage(newFriendEmail)
              setButtonTitle('ADD')
            }
          }
        }
      }
      getUserId()
    }, [newFriendID]);

    const searchNewFriend = async () => {
      try {
          let currentUserId = await SecureStore.getItemAsync('userId');

          const currentUserResult = await axios.get(userLink(currentUserId))

          setList(JSON.stringify(currentUserResult.data.friends))

          const result = await axios.get(usersLink())

          const data = result.data

          data.every(friend => {
            if (friend.email === newFriendEmail) {
              setdoesEmailExist(true);
              setNewFriendID(friend.id);
              return false; // Stop iteration if a match is found
            } else {
              setdoesEmailExist(false);
              setNewFriendID(null);
              return true; // Continue iteration if no match is found
            }
          });
      } catch (error) {
        console.log(error.message);
      }
    }

    const addNewFriend = async () => {
      try {
        const friendURL = userLink(newFriendID)
        const userURL = userLink(currentUserId)
        const result = await axios.get(userURL)

        let list = result.data.friends

        if (!list.includes(friendURL)) {
          list.push(friendURL);
        }
        
        await axios.patch(userURL,{
            'friends':list
        })

        const friendsRequests = list.map(url =>
          axios.get(url)
          .then(response => response.data)
        );
      
        Promise.all(friendsRequests)
          .then(friendsData => {
            if (JSON.stringify(friends) !== JSON.stringify(friendsData)) {
              setFriends(friendsData);
            }
          })
          .catch(error => {
            console.error('Error fetching friends:', error);
        });
      } catch (error) {
        console.log(error.message);
      }
    }


    useFocusEffect(
      React.useCallback(() => {
        return () => {
          const removeNotesFromReduxStore = async () => {
            await Promise.all(notes.map((sticker) => dispatch(removeNote(sticker.id))));
          };
          removeNotesFromReduxStore()
        };
      }, [notes])
    );

    return (
      <ScrollView>
        <TextInput placeholder='Insert friend email' onChangeText={(newFriendEmail) => setNewFriendEmail(newFriendEmail)} value={newFriendEmail}/>
        <Button title='SEARCH NEW FRIEND' onPress={()=>searchNewFriend()}/>
        <Text>{message}</Text>
        <Button title={buttonTitle} onPress={()=>addNewFriend()}/>
        <Text>Friends:</Text>
        <ScrollView>
          {friends.map((friend)=>(
            <Button key={friend.id} title={friend.email} onPress={() => navigation.navigate('FriendsBoard', { friendId: friend.id, friendName: friend.first_name })}/>
          ))}
        </ScrollView>
      </ScrollView>
    );
}

export default FriendsScreen