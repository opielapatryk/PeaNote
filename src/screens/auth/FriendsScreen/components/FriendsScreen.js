import { Text,View, Pressable, TextInput,FlatList } from 'react-native'
import React, { useEffect } from 'react'
import {findUser} from '../functions/findUser'
import {styles} from '../../../../../assets/styles/styles'
import auth from '@react-native-firebase/auth'
import {firebase} from '@react-native-firebase/database'
import { useDispatch, useSelector } from 'react-redux';
import {setEmail,setMessage} from '../../../../store/login/loginSlice'
import { useFocusEffect } from '@react-navigation/native';
import { loadUser } from '../functions/loadUser';
import { useSafeAreaInsets } from 'react-native-safe-area-context';
import { setFriendimage } from '../../../../store/settings/settingsSlice';
import { downloadImage } from '../../BoardScreen/functions/downloadImage';
import { cleanStoreFriends, removeFriendReducer } from '../../../../store/friends/friendsSlice'

export const FriendsScreen = ({ navigation }) => {
  const {friends} = useSelector((state) => state.friends);
  const {message,email} = useSelector((state) => state.login)
  const dispatch = useDispatch()
  const EMAIL = auth().currentUser.email

  useFocusEffect(
    React.useCallback(() => {
      loadUser(dispatch)
      const onChildAdd = () => loadUser(dispatch);
      const onChildRemove = () => {
        dispatch(cleanStoreFriends())
        loadUser(dispatch);
      };

      const listen = async ()=>{
        const usersRef = firebase.app().database('https://stickify-407810-default-rtdb.europe-west1.firebasedatabase.app/').ref('users')
        const snapshot = await usersRef.orderByChild('email').equalTo(EMAIL).once('value');
        const userData = snapshot.val();
        const userId = Object.keys(userData)[0];
      
        const friendsRef = firebase.app().database('https://stickify-407810-default-rtdb.europe-west1.firebasedatabase.app/').ref(`users/${userId}/friends`);

        friendsRef.on('child_changed', onChildAdd);
        friendsRef.on('child_removed', onChildRemove);
      }

      listen()
      return ()=>{
        const listenOff = async () => {
          const usersRef = firebase.app().database('https://stickify-407810-default-rtdb.europe-west1.firebasedatabase.app/').ref('users');
          const snapshot = await usersRef.orderByChild('email').equalTo(EMAIL).once('value');
          const userData = snapshot.val();
          const userId = Object.keys(userData)[0];
          const friendsRef = firebase.app().database('https://stickify-407810-default-rtdb.europe-west1.firebasedatabase.app/').ref(`users/${userId}/friends`);
  
          // Remove the 'child_added' listener when the component unmounts
          friendsRef.off('child_changed', onChildAdd);
          friendsRef.off('child_removed', onChildRemove);
        }

        listenOff()
        
        dispatch(setMessage(''))
        dispatch(setEmail(''))
      }
    }, [])
  );

  useEffect(()=>{
    loadUser(dispatch)
  },[])


  const renderFriends = ({ item }) => {
    return (
      <Pressable onPress={async () =>{
        await downloadImage(item.email).then((fileUri)=>{
          dispatch(setFriendimage(fileUri));
        })
        navigation.navigate('FriendsBoard', {name:item.username, friendEmail: item.email, oldnickname:item.nickname})
      }} style={styles.friendsList}><Text style={styles.firendListText}>{item.nickname?item.nickname:item.username}</Text></Pressable>
    );
  };

  const insets = useSafeAreaInsets();

  return (
    <View style={[styles.board,{
      paddingTop: insets.top,
      flex: 1,
      backgroundColor:'white',
    }]}>
      <Pressable style={styles.friendsHeaderRequest} onPress={()=>navigation.navigate('Requests')}><Text style={styles.removeFriendText}>Requests</Text></Pressable>

      <TextInput style={styles.friendsTextInput} placeholder={message?message:'Search by email or username'} onChangeText={text => dispatch(setEmail(text))} value={email} autoCorrect={false}/>

      <Pressable style={styles.friendsHeaderRequest} onPress={()=>{
        findUser(dispatch,email,navigation)}}><Text style={styles.removeFriendText}>Search</Text></Pressable>

      <FlatList data={friends} renderItem={({item})=>renderFriends({item})} keyExtractor={(friend) => friend.id}/>
    </View>
  );
}

export default FriendsScreen