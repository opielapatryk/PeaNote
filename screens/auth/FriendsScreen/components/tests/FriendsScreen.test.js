import React from 'react';
import renderer from 'react-test-renderer';
import FriendsScreen from '../FriendsScreen';
import { NavigationContainer } from '@react-navigation/native';
import { act } from 'react-test-renderer';

jest.mock('@react-native-firebase/auth', () => () => {
    return {
        currentUser: {
            email:'test@example.com'
        }
    }
  })

  jest.mock('@react-native-firebase/firestore', () =>()=> {
    return {
      collection: jest.fn(()=>{
        return{
          where: jest.fn(()=>{
            return{
              get: jest.fn(() => Promise.resolve(true))
            }
          }),
        }
      }),
    };
  })

test('FriendsBoard screen renders correctly.', async () => {
    const navigation = {
      navigate: jest.fn(),
    };

    const tree = renderer.create(
        <NavigationContainer>
            <FriendsScreen navigation={navigation} />
        </NavigationContainer>
      ).toJSON();

      await act(async () => {
        expect(tree).toMatchSnapshot();
      })
})
