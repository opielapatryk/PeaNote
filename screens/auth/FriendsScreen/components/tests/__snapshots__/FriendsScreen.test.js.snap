// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FriendsBoard screen renders correctly. 1`] = `
<View
  style={
    {
      "backgroundColor": "#FFFBE4",
      "flex": 1,
      "paddingTop": 10,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <Text
      style={
        {
          "alignSelf": "center",
          "backgroundColor": "#FFF3B2",
          "borderRadius": 3,
          "borderWidth": 2,
          "elevation": 1,
          "fontSize": 20,
          "fontWeight": "bold",
          "letterSpacing": 1,
          "margin": 10,
          "padding": 5,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 1,
            "width": 0,
          },
          "shadowOpacity": 0.18,
          "shadowRadius": 1,
        }
      }
    >
      REQUESTS
    </Text>
  </View>
  <TextInput
    onChangeText={[Function]}
    placeholder="Insert friend email"
    style={
      {
        "alignSelf": "center",
        "borderRadius": 5,
        "borderWidth": 1,
        "marginTop": 10,
        "padding": 5,
        "width": 300,
      }
    }
    value=""
  />
  <Text
    style={
      {
        "alignSelf": "center",
        "marginTop": 15,
      }
    }
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style=""
  >
    <Text
      style={
        {
          "height": 0,
        }
      }
    />
  </View>
  <Text
    style={
      {
        "alignSelf": "center",
        "fontSize": 20,
        "fontWeight": "bold",
        "marginTop": 15,
      }
    }
  >
    Friends:
  </Text>
  <RCTScrollView
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    style={
      {
        "margin": 10,
      }
    }
    viewabilityConfigCallbackPairs={[]}
  >
    <View />
  </RCTScrollView>
</View>
`;
